notifications/source/client.go

```diff
@@ -0,0 +1,187 @@
+// is needed to fetch the rules, as Icinga Notifications will respond with the newest ones whenever it detects that
+// you're using an outdated event rules config.
+//
+// If the request fails or the response is not as expected, it returns an error; otherwise, it returns nil.
+func (c *Client) ProcessEvent(ctx context.Context, ev *event.Event) (*RulesInfo, error) {
+	body, err := json.Marshal(ev)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot encode event to JSON")
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.Endpoints.ProcessEvent, bytes.NewReader(body))
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create HTTP request")
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Accept", "application/json")
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot POST HTTP request to process event")
+	}
+
+	defer func() {
+		_, _ = io.Copy(io.Discard, resp.Body)
+		_ = resp.Body.Close()
+	}()
+
+	if resp.StatusCode == http.StatusPreconditionFailed {
+		// Indicates that the rules version is outdated and the body should contain the current rules and their version.
+		// So, we read the body to extract the rules and return an ErrRulesOutdated error, so the caller can retry
+		// the event submission after it has reevaluated them.
+		var rulesInfo RulesInfo
+		if err := json.NewDecoder(resp.Body).Decode(&rulesInfo); err != nil {
+			return nil, errors.Wrap(err, "cannot decode new rules from process event response")
+		}
+
+		return &rulesInfo, ErrRulesOutdated
+	}
+
+	if resp.StatusCode >= http.StatusOK && resp.StatusCode <= 299 {
+		return nil, nil // Successfully processed the event.
+	}
+
+	if resp.StatusCode == http.StatusNotAcceptable {
+		return nil, nil // Superfluous state change event.
+	}
+
+	var buf bytes.Buffer
+	_, _ = io.Copy(&buf, &io.LimitedReader{R: resp.Body, N: 1 << 20}) // Limit the error message length to avoid memory exhaustion.
+
+	return nil, errors.Errorf("unexpected response from process event API, status %q (%d): %q",
+		resp.Status, resp.StatusCode, strings.TrimSpace(buf.String()))
+}
+
+// JoinIcingaWeb2Path constructs a URL by joining the Icinga Web 2 base URL with the provided relative URL.
+//
+// It is used to convert any relative URL into an absolute URL that points to the Icinga Web 2 instance.
+// A relative URL like "/icingadb/host" is transformed to e.g. "https://icinga.example.com/icingaweb2/icingadb/host"
+// after passing through this method, assuming the Icinga Web 2 base URL is "https://icinga.example.com/icingaweb2".
+func (c *Client) JoinIcingaWeb2Path(relativePath string) *url.URL {
+	return c.IcingaWebBaseUrl.JoinPath(relativePath)
+}
+
+// EmptyRulesVersion is a constant representing the version of the rules when no rules are present.
+// It is used to indicate that there are no rules available for a given source.
+const EmptyRulesVersion = "0x0"
```

Comment:
Do we need this? We shouldn't need to treat this specially on the source side, that should just see any opaque rule version, no matter if this describes the empty rule set or not. (We could treat the empty ruleset special on the daemon side though because it's empty, we know that the result of evaluation it will be that no rules match, so in that case, the event can be accepted without having the check the rule version.)

Reply:
No, this was mostly an artifact from prior versions of this PR. Removed.